# Exercise 02 -  OLAP Cubes - Grouping Sets

All the databases table in this demo are based on public database samples and transformations
- `Sakila` is a sample database created by `MySql` [Link](https://dev.mysql.com/doc/sakila/en/sakila-structure.html)
- The postgresql version of it is called `Pagila` [Link](https://github.com/devrimgunduz/pagila)
- The facts and dimension tables design is based on O'Reilly's public dimensional modelling tutorial schema [Link](http://archive.oreilly.com/oreillyschool/courses/dba3/index.html)

Start by connecting to the database by running the cells below. If you are coming back to this exercise, then uncomment and run the first cell to recreate the database. If you recently completed the slicing and dicing exercise, then skip to the second cell.

In [1]:
!PGPASSWORD=student createdb -h 127.0.0.1 -U student pagila_star
!PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila_star -f Data/pagila-star.sql

 set_config 
------------
 
(1 row)

 setval 
--------
    200
(1 row)

 setval 
--------
    605
(1 row)

 setval 
--------
     16
(1 row)

 setval 
--------
    600
(1 row)

 setval 
--------
    109
(1 row)

 setval 
--------
    599
(1 row)

 setval 
--------
      1
(1 row)

 setval 
--------
      1
(1 row)

 setval 
--------
      1
(1 row)

 setval 
--------
      1
(1 row)

 setval 
--------
  16049
(1 row)

 setval 
--------
   1000
(1 row)

 setval 
--------
   4581
(1 row)

 setval 
--------
      6
(1 row)

 setval 
--------
  32098
(1 row)

 setval 
--------
  16049
(1 row)

 setval 
--------
      2
(1 row)

 setval 
--------
      2
(1 row)



### Connect to the local database where Pagila is loaded

In [2]:
import sql
%load_ext sql

DB_ENDPOINT = "127.0.0.1"
DB = 'pagila_star'
DB_USER = 'student'
DB_PASSWORD = 'student'
DB_PORT = '5432'

# postgresql://username:password@host:port/database
conn_string = "postgresql://{}:{}@{}:{}/{}" \
                        .format(DB_USER, DB_PASSWORD, DB_ENDPOINT, DB_PORT, DB)

print(conn_string)
%sql $conn_string

postgresql://student:student@127.0.0.1:5432/pagila_star


'Connected: student@pagila_star'

### Star Schema

<img src="pagila-star.png" width="50%"/>

# Grouping Sets
- It happens often that for 3 dimensions, you want to aggregate a fact:
    - by nothing (total)
    - then by the 1st dimension
    - then by the 2nd 
    - then by the 3rd 
    - then by the 1st and 2nd
    - then by the 2nd and 3rd
    - then by the 1st and 3rd
    - then by the 1st and 2nd and 3rd
    
- Since this is very common, and in all cases, we are iterating through all the fact table anyhow, there is a more clever way to do that using the SQL grouping statement "GROUPING SETS" 

## Total Revenue

TODO: Write a query that calculates total revenue (sales_amount)

In [3]:
%%sql
SELECT sum(sales_amount) revenue
FROM factsales;

 * postgresql://student:***@127.0.0.1:5432/pagila_star
1 rows affected.


[(Decimal('67416.51'),)]

## Revenue by Country
TODO: Write a query that calculates total revenue (sales_amount) by country

In [6]:
%%sql
SELECT c.country, sum(s.sales_amount) as revenue
FROM factsales s
JOIN dimcustomer c ON (s.customer_key = c.customer_key)
GROUP BY country
ORDER BY revenue desc;

 * postgresql://student:***@127.0.0.1:5432/pagila_star
108 rows affected.


[('India', Decimal('6630.27')),
 ('China', Decimal('5802.73')),
 ('United States', Decimal('4110.32')),
 ('Japan', Decimal('3471.74')),
 ('Mexico', Decimal('3307.04')),
 ('Brazil', Decimal('3200.52')),
 ('Russian Federation', Decimal('3045.87')),
 ('Philippines', Decimal('2381.32')),
 ('Turkey', Decimal('1662.12')),
 ('Nigeria', Decimal('1511.48')),
 ('Indonesia', Decimal('1510.33')),
 ('Argentina', Decimal('1434.48')),
 ('Taiwan', Decimal('1210.94')),
 ('South Africa', Decimal('1204.15')),
 ('Iran', Decimal('950.75')),
 ('United Kingdom', Decimal('924.80')),
 ('Poland', Decimal('877.97')),
 ('Italy', Decimal('831.11')),
 ('Germany', Decimal('831.04')),
 ('Vietnam', Decimal('746.28')),
 ('Ukraine', Decimal('730.42')),
 ('Colombia', Decimal('709.41')),
 ('Egypt', Decimal('694.39')),
 ('Venezuela', Decimal('683.30')),
 ('Spain', Decimal('606.58')),
 ('Canada', Decimal('593.63')),
 ('Netherlands', Decimal('586.66')),
 ('South Korea', Decimal('574.65')),
 ('Pakistan', Decimal('525.72')),
 ('Saudi Arabia', Decimal('523.79')),
 ('Yemen', Decimal('510.83')),
 ('Peru', Decimal('468.88')),
 ('Israel', Decimal('422.01')),
 ('Thailand', Decimal('419.04')),
 ('Bangladesh', Decimal('402.05')),
 ('Ecuador', Decimal('390.13')),
 ('Algeria', Decimal('383.10')),
 ('France', Decimal('374.04')),
 ('Malaysia', Decimal('369.15')),
 ('Mozambique', Decimal('344.20')),
 ('Tanzania', Decimal('341.17')),
 ('United Arab Emirates', Decimal('333.16')),
 ('Chile', Decimal('328.29')),
 ('Dominican Republic', Decimal('318.23')),
 ('Morocco', Decimal('307.29')),
 ('Austria', Decimal('307.22')),
 ('Belarus', Decimal('277.34')),
 ('Paraguay', Decimal('275.38')),
 ('Latvia', Decimal('262.40')),
 ('Yugoslavia', Decimal('259.43')),
 ('Puerto Rico', Decimal('254.39')),
 ('Kenya', Decimal('253.46')),
 ('Switzerland', Decimal('252.39')),
 ('Azerbaijan', Decimal('245.43')),
 ('French Polynesia', Decimal('235.46')),
 ('Romania', Decimal('235.38')),
 ('Greece', Decimal('232.46')),
 ('Sudan', Decimal('227.46')),
 ('Runion', Decimal('216.54')),
 ('Angola', Decimal('215.48')),
 ('Congo, The Democratic Republic of the', Decimal('212.50')),
 ('Bulgaria', Decimal('204.50')),
 ('Cameroon', Decimal('200.46')),
 ('Kazakstan', Decimal('198.48')),
 ('Myanmar', Decimal('194.48')),
 ('Cambodia', Decimal('191.47')),
 ('Oman', Decimal('187.50')),
 ('Bolivia', Decimal('183.53')),
 ('Holy See (Vatican City State)', Decimal('152.66')),
 ('Nauru', Decimal('148.69')),
 ('Sweden', Decimal('144.66')),
 ('Hong Kong', Decimal('142.70')),
 ('Greenland', Decimal('137.66')),
 ('Turkmenistan', Decimal('136.73')),
 ('Chad', Decimal('135.68')),
 ('Zambia', Decimal('134.67')),
 ('Czech Republic', Decimal('133.71')),
 ('Gambia', Decimal('129.70')),
 ('Moldova', Decimal('127.66')),
 ('Malawi', Decimal('123.72')),
 ('Virgin Islands, U.S.', Decimal('122.68')),
 ('Tuvalu', Decimal('120.74')),
 ('Armenia', Decimal('118.75')),
 ('Nepal', Decimal('116.78')),
 ('Sri Lanka', Decimal('116.70')),
 ('Estonia', Decimal('115.70')),
 ('Liechtenstein', Decimal('114.72')),
 ('Faroe Islands', Decimal('114.72')),
 ('North Korea', Decimal('113.69')),
 ('Brunei', Decimal('113.65')),
 ('Bahrain', Decimal('112.75')),
 ('Kuwait', Decimal('111.74')),
 ('Iraq', Decimal('111.73')),
 ('Hungary', Decimal('111.71')),
 ('Anguilla', Decimal('106.65')),
 ('French Guiana', Decimal('103.78')),
 ('Finland', Decimal('101.74')),
 ('Senegal', Decimal('100.75')),
 ('Saint Vincent and the Grenadines', Decimal('96.75')),
 ('Madagascar', Decimal('93.78')),
 ('New Zealand', Decimal('92.76')),
 ('Ethiopia', Decimal('91.77')),
 ('Slovakia', Decimal('89.74')),
 ('Tunisia', Decimal('78.77')),
 ('Tonga', Decimal('73.82')),
 ('Lithuania', Decimal('73.76')),
 ('American Samoa', Decimal('71.80')),
 ('Afghanistan', Decimal('67.82'))]

## Revenue by Month
TODO: Write a query that calculates total revenue (sales_amount) by month

In [7]:
%%sql
SELECT d.month, sum(s.sales_amount) AS revenue
FROM factsales s
JOIN dimdate d ON (s.date_key = d.date_key)
GROUP BY month
ORDER BY revenue desc;

 * postgresql://student:***@127.0.0.1:5432/pagila_star
5 rows affected.


[(4, Decimal('28559.46')),
 (3, Decimal('23886.56')),
 (2, Decimal('9631.88')),
 (1, Decimal('4824.43')),
 (5, Decimal('514.18'))]

## Revenue by Month & Country
TODO: Write a query that calculates total revenue (sales_amount) by month and country. Sort the data by month, country, and revenue in descending order. The first few rows of your output should match the table below.

In [12]:
%%sql
SELECT d.month, c.country, sum(s.sales_amount) AS revenue
FROM factsales s
JOIN dimdate d ON (s.date_key = d.date_key)
JOIN dimcustomer c ON (s.customer_key = c.customer_key)
GROUP BY (d.month, c.country, s.sales_amount)
ORDER BY (d.month, c.country, s.sales_amount) DESC
LIMIT 20;

 * postgresql://student:***@127.0.0.1:5432/pagila_star
2768 rows affected.


[(5, 'Zambia', Decimal('0.99')),
 (5, 'Yemen', Decimal('2.99')),
 (5, 'Vietnam', Decimal('4.99')),
 (5, 'Vietnam', Decimal('2.99')),
 (5, 'Vietnam', Decimal('0.99')),
 (5, 'Venezuela', Decimal('5.98')),
 (5, 'United States', Decimal('9.98')),
 (5, 'United States', Decimal('7.96')),
 (5, 'United States', Decimal('8.97')),
 (5, 'United States', Decimal('1.98')),
 (5, 'United States', Decimal('0.00')),
 (5, 'United Kingdom', Decimal('9.98')),
 (5, 'United Kingdom', Decimal('1.98')),
 (5, 'United Arab Emirates', Decimal('4.99')),
 (5, 'Turkmenistan', Decimal('2.99')),
 (5, 'Turkey', Decimal('4.99')),
 (5, 'Turkey', Decimal('5.98')),
 (5, 'Tunisia', Decimal('0.99')),
 (5, 'Thailand', Decimal('5.98')),
 (5, 'Thailand', Decimal('0.00')),
 (5, 'Tanzania', Decimal('4.99')),
 (5, 'Taiwan', Decimal('8.97')),
 (5, 'Taiwan', Decimal('5.98')),
 (5, 'Taiwan', Decimal('2.99')),
 (5, 'Taiwan', Decimal('0.99')),
 (5, 'Taiwan', Decimal('0.00')),
 (5, 'Switzerland', Decimal('7.98')),
 (5, 'Switzerland', Decimal('0.00')),
 (5, 'Spain', Decimal('9.98')),
 (5, 'South Korea', Decimal('2.99')),
 (5, 'South Korea', Decimal('0.99')),
 (5, 'South Africa', Decimal('4.99')),
 (5, 'Saudi Arabia', Decimal('2.99')),
 (5, 'Saudi Arabia', Decimal('0.99')),
 (5, 'Russian Federation', Decimal('4.99')),
 (5, 'Russian Federation', Decimal('2.97')),
 (5, 'Puerto Rico', Decimal('0.99')),
 (5, 'Philippines', Decimal('14.97')),
 (5, 'Philippines', Decimal('5.98')),
 (5, 'Philippines', Decimal('4.95')),
 (5, 'Peru', Decimal('4.99')),
 (5, 'Oman', Decimal('4.99')),
 (5, 'Nigeria', Decimal('5.98')),
 (5, 'Nigeria', Decimal('2.99')),
 (5, 'Nigeria', Decimal('0.00')),
 (5, 'Netherlands', Decimal('0.99')),
 (5, 'Morocco', Decimal('4.99')),
 (5, 'Mexico', Decimal('4.99')),
 (5, 'Mexico', Decimal('5.98')),
 (5, 'Mexico', Decimal('2.97')),
 (5, 'Malaysia', Decimal('0.99')),
 (5, 'Lithuania', Decimal('4.99')),
 (5, 'Latvia', Decimal('2.99')),
 (5, 'Kuwait', Decimal('0.99')),
 (5, 'Kenya', Decimal('2.99')),
 (5, 'Japan', Decimal('15.96')),
 (5, 'Japan', Decimal('2.99')),
 (5, 'Japan', Decimal('3.96')),
 (5, 'Japan', Decimal('0.00')),
 (5, 'Italy', Decimal('2.99')),
 (5, 'Italy', Decimal('0.99')),
 (5, 'Israel', Decimal('2.99')),
 (5, 'Iraq', Decimal('0.99')),
 (5, 'Iran', Decimal('4.99')),
 (5, 'Indonesia', Decimal('4.99')),
 (5, 'Indonesia', Decimal('0.99')),
 (5, 'India', Decimal('9.98')),
 (5, 'India', Decimal('5.98')),
 (5, 'India', Decimal('19.96')),
 (5, 'India', Decimal('7.96')),
 (5, 'India', Decimal('11.96')),
 (5, 'India', Decimal('6.93')),
 (5, 'India', Decimal('0.00')),
 (5, 'Hungary', Decimal('0.99')),
 (5, 'Greece', Decimal('0.99')),
 (5, 'Germany', Decimal('9.98')),
 (5, 'Germany', Decimal('3.98')),
 (5, 'Germany', Decimal('0.99')),
 (5, 'Germany', Decimal('0.00')),
 (5, 'French Polynesia', Decimal('4.99')),
 (5, 'French Polynesia', Decimal('3.98')),
 (5, 'French Polynesia', Decimal('0.00')),
 (5, 'France', Decimal('0.99')),
 (5, 'Faroe Islands', Decimal('0.99')),
 (5, 'Colombia', Decimal('14.97')),
 (5, 'China', Decimal('7.98')),
 (5, 'China', Decimal('14.97')),
 (5, 'China', Decimal('14.95')),
 (5, 'China', Decimal('2.97')),
 (5, 'China', Decimal('0.00')),
 (5, 'Canada', Decimal('2.99')),
 (5, 'Canada', Decimal('0.99')),
 (5, 'Cameroon', Decimal('3.98')),
 (5, 'Cameroon', Decimal('2.99')),
 (5, 'Cameroon', Decimal('0.00')),
 (5, 'Cambodia', Decimal('5.98')),
 (5, 'Cambodia', Decimal('3.98')),
 (5, 'Cambodia', Decimal('0.00')),
 (5, 'Bulgaria', Decimal('0.99')),
 (5, 'Brunei', Decimal('4.99')),
 (5, 'Brazil', Decimal('7.98')),
 (5, 'Brazil', Decimal('5.98')),
 (5, 'Brazil', Decimal('4.99')),
 (5, 'Brazil', Decimal('2.99')),
 (5, 'Brazil', Decimal('0.99')),
 (5, 'Brazil', Decimal('0.00')),
 (5, 'Bolivia', Decimal('2.99')),
 (5, 'Belarus', Decimal('2.99')),
 (5, 'Bangladesh', Decimal('2.99')),
 (5, 'Bahrain', Decimal('2.99')),
 (5, 'Azerbaijan', Decimal('0.99')),
 (5, 'Armenia', Decimal('2.99')),
 (5, 'Argentina', Decimal('5.98')),
 (5, 'Argentina', Decimal('14.97')

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>month</th>
        <th>country</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>1</td>
        <td>Australia</td>
        <td>2364.19</td>
    </tr>
    <tr>
        <td>1</td>
        <td>Canada</td>
        <td>2460.24</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Australia</td>
        <td>4895.10</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Canada</td>
        <td>4736.78</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Australia</td>
        <td>12060.33</td>
    </tr>
</tbody></table></div>

## Revenue Total, by Month, by Country, by Month & Country All in one shot

TODO: Write a query that calculates total revenue at the various grouping levels done above (total, by month, by country, by month & country) all at once using the grouping sets function. Your output should match the table below.

In [ ]:
%%sql
SELECT ...
FROM ...

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>month</th>
        <th>country</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>1</td>
        <td>Australia</td>
        <td>2364.19</td>
    </tr>
    <tr>
        <td>1</td>
        <td>Canada</td>
        <td>2460.24</td>
    </tr>
    <tr>
        <td>1</td>
        <td>None</td>
        <td>4824.43</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Australia</td>
        <td>4895.10</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Canada</td>
        <td>4736.78</td>
    </tr>
    <tr>
        <td>2</td>
        <td>None</td>
        <td>9631.88</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Australia</td>
        <td>12060.33</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Canada</td>
        <td>11826.23</td>
    </tr>
    <tr>
        <td>3</td>
        <td>None</td>
        <td>23886.56</td>
    </tr>
    <tr>
        <td>4</td>
        <td>Australia</td>
        <td>14136.07</td>
    </tr>
    <tr>
        <td>4</td>
        <td>Canada</td>
        <td>14423.39</td>
    </tr>
    <tr>
        <td>4</td>
        <td>None</td>
        <td>28559.46</td>
    </tr>
    <tr>
        <td>5</td>
        <td>Australia</td>
        <td>271.08</td>
    </tr>
    <tr>
        <td>5</td>
        <td>Canada</td>
        <td>243.10</td>
    </tr>
    <tr>
        <td>5</td>
        <td>None</td>
        <td>514.18</td>
    </tr>
    <tr>
        <td>None</td>
        <td>None</td>
        <td>67416.51</td>
    </tr>
    <tr>
        <td>None</td>
        <td>Australia</td>
        <td>33726.77</td>
    </tr>
    <tr>
        <td>None</td>
        <td>Canada</td>
        <td>33689.74</td>
    </tr>
</tbody></table></div>